src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -0,0 +1,774 @@
+            _ = !string.IsNullOrEmpty(userTable.FullName) ? true : throw new ArgumentNullException(nameof(userTable));
+            _ = !string.IsNullOrEmpty(userFunctionId) ? true : throw new ArgumentNullException(nameof(userFunctionId));
+            _ = !string.IsNullOrEmpty(workerTableName) ? true : throw new ArgumentNullException(nameof(workerTableName));
+            _ = userTableColumns ?? throw new ArgumentNullException(nameof(userTableColumns));
+            _ = primaryKeyColumns ?? throw new ArgumentNullException(nameof(primaryKeyColumns));
+            _ = executor ?? throw new ArgumentNullException(nameof(executor));
+            _ = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            this._connectionString = connectionString;
+            this._userTableId = userTableId;
+            this._userTable = userTable;
+            this._userFunctionId = userFunctionId;
+            this._workerTableName = workerTableName;
+            this._userTableColumns = primaryKeyColumns.Concat(userTableColumns.Except(primaryKeyColumns)).ToList();
+            this._primaryKeyColumns = primaryKeyColumns;
+
+            // Prep search-conditions that will be used besides WHERE clause to match table rows.
+            this._rowMatchConditions = Enumerable.Range(0, BatchSize)
+                .Select(index => string.Join(" AND ", primaryKeyColumns.Select(col => $"{col} = @{col}_{index}")))
+                .ToList();
+
+            this._executor = executor;
+            this._logger = logger;
+
+            this._cancellationTokenSourceCheckForChanges = new CancellationTokenSource();
+            this._cancellationTokenSourceRenewLeases = new CancellationTokenSource();
+            this._cancellationTokenSourceExecutor = new CancellationTokenSource();
+
+            this._rowsLock = new SemaphoreSlim(1);
+            this._rows = new List<IReadOnlyDictionary<string, string>>();
+            this._leaseRenewalCount = 0;
+            this._state = State.CheckingForChanges;
+
+#pragma warning disable CS4014 // Queue the below tasks and exit. Do not wait for their completion.
+            _ = Task.Run(() =>
+            {
+                this.RunChangeConsumptionLoopAsync();
+                this.RunLeaseRenewalLoopAsync();
+            });
+#pragma warning restore CS4014
+        }
+
+        /// <summary>
+        /// Stops the change monitor which stops polling for changes on the user's table. If the change monitor is
+        /// currently executing a set of changes, it is only stopped once execution is finished and the user's function
+        /// is triggered (whether or not the trigger is successful).
+        /// </summary>
+        public void Stop()
+        {
+            this._cancellationTokenSourceCheckForChanges.Cancel();
+        }
+
+        public void Dispose()
+        {
+            throw new NotImplementedException();
```

Comment:
Is this needed? And shouldn't dispose be calling stop?

Reply:
This piece of code must be added since the class `SqlTableChangeMonitor` otherwise gets flagged by CodeAnalysis as it owns disposable resources of type `CancellationTokenSource` but does not implement `IDisposable` itself. I have replaced `Stop()` method with `Dispose()` for now as there is no requirement posed by WebJobs SDK for change-monitor class to have method specifically with same `Stop`. Please reopen the conversation if you have a better solution for this. Other way is to add suppression for the warning message or surround the code by `#pragma disable` directive, but I see at no place in the rest of project are we suppressing messages other than 'unused parameters'.

Reply:
A few notes: In general, most other extensions have the logic that listens to new events entirely encoded inside the `*Listener` class. In SQL extension, we have the main `SqlTriggerListener` class and its extension `SqlTableChangeMonitor` class. The reason is that we require a couple of variables (e.g., `userTableId') that need to be initialized after the listener is started. But these variables then need to be set to mutable which will get error-prone considering that the listener logic contains concurrency. So, we kept all one-time initialization steps inside `SqlTriggerListener`, got the local variables initialized, and passed them to `SqlTableChangeMonitor` which stores them in read-only variables.

Earlier we started having similar method definitions inside both classes i.e., `StartAsync` and `StopAsync` but since `StartAsync` calls `SqlTableChangeMonitor`'s constructor, having `StopAsync` call it's `Dispose` method kinds of feels consistent to me.

By the way, I will be updating the `Dispose` method in outer `SqlTriggerListener` class to do nothing like in [here](https://github.com/Azure/azure-webjobs-sdk-extensions/blob/48fe14f0a373a3a279849eae0cf0b3756fa3d47f/src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerListener.cs#L85-L88) as from the implementation, it seems that WebJobs SDK would be always call either of `StopAsync` or `Cancel` or both.